### DISCUSSION TITLE
How to apply limit to the whole query instead of the internal ?

### DISCUSSION THREAD

**Discussion by jossefaz** [sqlalchemy/sqlalchemy#8043] [Usage Questions]
I have three models
```python
class Model1(Base):
    __tablename__ = 'model1'
    id = Column(Integer, primary_key=True)
    name =  Column(String(50))
    rel_2 = relationship("Model2")
    rel_3 = relationship("Model3")

class Model2(Base):
    __tablename__ = 'model2'
    id = Column(Integer, primary_key=True)
    name =  Column(String(50))
    model1_id = Column (Integer,   ForeignKey('model1.id'))
    

class Model3(Base):
    __tablename__ = 'model3'
    id = Column(Integer, primary_key=True)
    name =  Column(String(50)) 
    model1_id = Column (Integer,   ForeignKey('model1.id'))
```
I am trying to select from Model1.

So far I have 

```python
stmt = select(Model1)
           .join(Model2)
           .join(Model3)        
           .options(contains_eager(Model1.rel_2),
                         contains_eager(Model1.rel_3)
           .where(model2.name == 'hello')
           .offset(0).limit(100)
```
This will give me this kind of query
```sql
SELECT anon_1.id, anon_1.name
FROM (SELECT model1.id AS id, 
             model1.name as name
             model2.id as id_1
             model2.name as name_1
             model3.id as id_2
             model3.name as name_2
      FROM model1
               INNER JOIN model2 ON model1.id = model2.model1_id
               INNER JOIN model3 ON model1.id = model3.model1_id
      WHERE model2.name = 'hello'
      LIMIT 0, 100) AS anon_1
```

The problem is the limit here. I need a pagination of 100 `model1` instances. But because of the joins I got lot of rows and limit 100 will return a random number of `model1` instances since this limit applies to the JOINS (and it could be hundreds of rows.

How to keep my paginations related to `model1` instances number ? I know it is more a pure SQL question but also I would like to know how to do achieve this with the ORM...

**Comment by jossefaz** [comment]
What I did so far is to apply a distinct on ids from model1 and only then perform the joins. It works but I wonder if there is a better way

**Reply by CaselIT** [reply]
Not sure what are you using the query for, have you looked at using `selectinload` instead of the explicit joins?

**Comment by zzzeek** [comment]
> The problem is the limit here. I need a pagination of 100 `model1` instances. But because of the joins I got lot of rows and limit 100 will return a random number of `model1` instances since this limit applies to the JOINS (and it could be hundreds of rows.

In SQL if you want to limit the inner rows and then join you use a subquery.  Since you are trying to select Model1 on criteria that's on Model2, you also need to fix that up so that you can get the Model1 with one row per Model1 primary key.   it's an awkward situation in that regard. SQL like this could work


```sql

SELECT subq.*, model2.*, model3.* FROM
(SELECT * FROM model1 WHERE EXISTS (SELECT 1 FROM model2 WHERE model2.model1_id = model1.id AND model2.name=='hello') LIMIT 100) AS subq JOIN model2 ON ... JOIN model3 ON ...  WHERE model2.name == 'hello'

```

```py
subq = (
    select(Model1).
    where(exists().where(Model2.model1_id == Model1.id, Model2.name == 'hello')).
    limit(100).
    subquery()
)

a1 = aliased(Model1, subq)
stmt = select(a1).join(Model2).join(Model3).options(
   contains_eager(a1.rel_2),
   contains_eager(a1.rel_3)
).where(Model2.name == 'hello')

```
